flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/sink/PubSubWriter.java

```diff
@@ -0,0 +1,195 @@
+
+    /** The PubSub generic client to publish messages. */
+    private final GcpWriterClient<PubsubMessage> publisher;
+
+    /**
+     * The maximum number of inflight requests, The writer blocks new writes if the number of
+     * inflight requests exceeds the specified limit.
+     */
+    private final long maximumInflightRequests;
+
+    /**
+     * Flag to indicate whether to fail on errors, if unset the writer will retry non-fatal request
+     * failures.
+     */
+    private final boolean failOnError;
+
+    private long inflightRequests = 0;
+
+    private final MailboxExecutor mailboxExecutor;
+    private final Counter numBytesOutCounter;
+    private final Counter numRecordsOutCounter;
+    private final Counter numRecordsOutErrorCounter;
+    private final SerializationSchema<T> serializationSchema;
+
+    PubSubWriter(
+            String projectId,
+            String topicId,
+            SerializationSchema<T> serializationSchema,
+            WriterInitContext context,
+            GcpPublisherConfig publisherConfig,
+            long maximumInflightRequests,
+            boolean failOnError)
+            throws IOException {
+        this(
+                createPublisher(projectId, topicId, publisherConfig),
+                context,
+                serializationSchema,
+                maximumInflightRequests,
+                failOnError);
+    }
+
+    @VisibleForTesting
+    PubSubWriter(
+            GcpWriterClient<PubsubMessage> publisher,
+            WriterInitContext context,
+            SerializationSchema<T> serializationSchema,
+            long maximumInflightRequests,
+            boolean failOnError) {
+        this.publisher = Preconditions.checkNotNull(publisher);
+        this.serializationSchema = Preconditions.checkNotNull(serializationSchema);
+        Preconditions.checkNotNull(context, "Context cannot be null.");
+
+        this.mailboxExecutor = context.getMailboxExecutor();
+        this.numBytesOutCounter = context.metricGroup().getIOMetricGroup().getNumBytesOutCounter();
+        this.numRecordsOutCounter =
+                context.metricGroup().getIOMetricGroup().getNumRecordsOutCounter();
+        this.numRecordsOutErrorCounter = context.metricGroup().getNumRecordsOutErrorsCounter();
+        this.maximumInflightRequests = maximumInflightRequests;
+        this.failOnError = failOnError;
+    }
+
+    @Override
+    public void write(T t, SinkWriter.Context context) throws IOException, InterruptedException {
+        awaitMaxInflightRequestsBelow(maximumInflightRequests);
+        PubsubMessage message =
+                PubsubMessage.newBuilder()
+                        .setData(ByteString.copyFrom(serializationSchema.serialize(t)))
+                        .build();
+        publishMessage(message);
+    }
+
+    @Override
+    public void flush(boolean b) throws IOException, InterruptedException {
+        publisher.flush();
+        awaitMaxInflightRequestsBelow(1);
```

Comment:
@vahmed-hamdy why do you need this additional `yield` if any in-flight message? Can you clarify?

If not obvious (and I am missing something), could you put a comment?

Reply:
@affo the inflight requests callback is queued in the mailbox, we yield here for callback executions, shall requests be sent successfully the call back should reset the inflight requests back to 0 exiting the loop